Allow `DatasetListItem` in `Client.delete_dataset`

https://app.circleci.com/pipelines/github/tswast/python-bigquery-pandas/217/workflows/38a35d17-9e68-48b3-b17a-b683e4ab5a52/jobs/1308

I expected this code to work:

```python
    @pytest.fixture(scope="session", autouse=True)
    def cleanup_datasets(bigquery_client: bigquery.Client):
        for dataset in bigquery_client.list_datasets():
            if prefixer.should_cleanup(dataset.dataset_id):
                bigquery_client.delete_dataset(
>                   dataset, delete_contents=True, not_found_ok=True
                )
```

But I get

```
if not isinstance(dataset, (Dataset, DatasetReference)):
>           raise TypeError("dataset must be a Dataset or a DatasetReference")
E           TypeError: dataset must be a Dataset or a DatasetReference

/opt/conda/envs/test-environment/lib/python3.7/site-packages/google/cloud/bigquery/client.py:740: TypeError
```

Similarly, we should check `delete_table`. I think it accepts `TableListItem`, but this is not reflected in the type annotations.

--- a/google/cloud/bigquery/client.py
+++ b/google/cloud/bigquery/client.py
@@ -555,7 +555,7 @@ class Client(ClientWithProject):
 
     def create_dataset(
         self,
-        dataset: Union[str, Dataset, DatasetReference],
+        dataset: Union[str, Dataset, DatasetReference, DatasetListItem],
         exists_ok: bool = False,
         retry: retries.Retry = DEFAULT_RETRY,
         timeout: float = DEFAULT_TIMEOUT,
@@ -685,7 +685,7 @@ class Client(ClientWithProject):
 
     def create_table(
         self,
-        table: Union[str, Table, TableReference],
+        table: Union[str, Table, TableReference, TableListItem],
         exists_ok: bool = False,
         retry: retries.Retry = DEFAULT_RETRY,
         timeout: float = DEFAULT_TIMEOUT,
@@ -699,6 +699,7 @@ class Client(ClientWithProject):
             table (Union[ \
                 google.cloud.bigquery.table.Table, \
                 google.cloud.bigquery.table.TableReference, \
+                google.cloud.bigquery.table.TableListItem, \
                 str, \
             ]):
                 A :class:`~google.cloud.bigquery.table.Table` to create.
@@ -1301,7 +1302,7 @@ class Client(ClientWithProject):
 
     def list_models(
         self,
-        dataset: Union[Dataset, DatasetReference, str],
+        dataset: Union[Dataset, DatasetReference, DatasetListItem, str],
         max_results: int = None,
         page_token: str = None,
         retry: retries.Retry = DEFAULT_RETRY,
@@ -1378,7 +1379,7 @@ class Client(ClientWithProject):
 
     def list_routines(
         self,
-        dataset: Union[Dataset, DatasetReference, str],
+        dataset: Union[Dataset, DatasetReference, DatasetListItem, str],
         max_results: int = None,
         page_token: str = None,
         retry: retries.Retry = DEFAULT_RETRY,
@@ -1455,7 +1456,7 @@ class Client(ClientWithProject):
 
     def list_tables(
         self,
-        dataset: Union[Dataset, DatasetReference, str],
+        dataset: Union[Dataset, DatasetReference, DatasetListItem, str],
         max_results: int = None,
         page_token: str = None,
         retry: retries.Retry = DEFAULT_RETRY,
@@ -1531,7 +1532,7 @@ class Client(ClientWithProject):
 
     def delete_dataset(
         self,
-        dataset: Union[Dataset, DatasetReference, str],
+        dataset: Union[Dataset, DatasetReference, DatasetListItem, str],
         delete_contents: bool = False,
         retry: retries.Retry = DEFAULT_RETRY,
         timeout: float = DEFAULT_TIMEOUT,

